web/src/components/areagraph.js

```diff
@@ -1,466 +1,278 @@
-                .attr('cx', this.x(this._data[this._selectedIndex].datetime))
-                .attr('cy', this.y(selectedData[1]))
-                .style('display', 'block')
-                .style('fill', function() {
-                    return fillColor.call(that, {
-                        key: that.stackKeys[that._selectedLayerIndex]
-                    })
-                })
-        } else {
-            this.markerElement.style('display', 'none')
-        }
+    if (mouseMoveHandler) {
+      mouseMoveHandler(layer[i].data._countryData, i);
     }
-    return this;
-}
-AreaGraph.prototype.displayByEmissions = function(arg) {
-    if (arg === undefined) return this._displayByEmissions;
-    else {
-        this._displayByEmissions = arg;
-        // Quick hack to re-render
-        // TODO: In principle we shouldn't be calling `.data()`
-        this
-            .data(this._originalData)
-            .render();
+  };
+  const handleLayerMouseOut = () => {
+    mouseOutTimeout = setTimeout(() => {
+      setSelectedLayerIndex(undefined);
+      if (mouseOutHandler) {
+        mouseOutHandler();
+      }
+      if (layerMouseOutHandler) {
+        layerMouseOutHandler();
+      }
+    }, 50);
+  };
+
+  return (
+    <g>
+      {stackedData.map((layer, ind) => (
+        <path
+          key={stackKeys[ind]}
+          className={`area layer ${stackKeys[ind]}`}
+          fill={fillColor(stackKeys[ind], displayByEmissions)}
+          style={{ cursor: 'pointer' }}
+          d={area(layer)}
+          onFocus={ev => handleLayerMouseMove(ev, layer, ind)}
+          onMouseOver={ev => handleLayerMouseMove(ev, layer, ind)}
+          onMouseMove={ev => handleLayerMouseMove(ev, layer, ind)}
+          onMouseOut={handleLayerMouseOut}
+          onBlur={handleLayerMouseOut}
+        />
+      ))}
+    </g>
+  );
+});
+
+const getCurrentTime = state =>
+  state.application.customDate || (state.data.grid || {}).datetime;
+
+// Regular production mode or exchange fill as a fallback
+const fillColor = (key, displayByEmissions) =>
+  modeColor[key] || (displayByEmissions ? 'darkgray' : `url(#areagraph-exchange-${key})`);
+
+const mapStateToProps = (state, props) => ({
+  colorBlindModeEnabled: state.application.colorBlindModeEnabled,
+  currentTime: getCurrentTime(state),
+  data: props.dataSelector(state),
+  displayByEmissions: state.application.tableDisplayEmissions,
+  electricityMixMode: state.application.electricityMixMode,
+  selectedIndex: state.application.selectedZoneTimeIndex,
+});
+
+const AreaGraph = ({
+  colorBlindModeEnabled,
+  currentTime,
+  data,
+  displayByEmissions,
+  electricityMixMode,
+  id,
+  selectedIndex,
+  layerMouseMoveHandler,
+  layerMouseOutHandler,
+  mouseMoveHandler,
+  mouseOutHandler,
+}) => {
+  // Hack to create a reference via non-mutable state so that it's created
+  // only once at the initialization (like in a class constructor).
+  const [ref] = useState(React.createRef());
```

Comment:
can't we use https://reactjs.org/docs/hooks-reference.html#useref?

Reply:
Yes of course, good catch, not sure why it didn't cross my mind :D